python-checks/src/main/java/org/sonar/python/checks/GenericFunctionTypeParameterCheck.java

```diff
@@ -0,0 +1,108 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.PythonVersionUtils;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.FunctionDef;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.Parameter;
+import org.sonar.plugins.python.api.tree.ParameterList;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.plugins.python.api.tree.TypeAnnotation;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6796")
+public class GenericFunctionTypeParameterCheck extends PythonSubscriptionCheck {
+
+  private static final String MESSAGE = "Use a generic type parameter for this function instead of a \"TypeVar\".";
+  private static final String SECONDARY_MESSAGE_USE = "Use of \"TypeVar\" here.";
+  private static final String SECONDARY_MESSAGE_ASSIGNMENT = "\"TypeVar\" is assigned here.";
+  public static final String TYPE_VAR_FQN = "typing.TypeVar";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FUNCDEF, GenericFunctionTypeParameterCheck::checkUseOfGenerics);
+  }
+
+  private static void checkUseOfGenerics(SubscriptionContext ctx) {
+    if (!supportsTypeParameterSyntax(ctx)) {
+      return;
+    }
+    FunctionDef functionDef = (FunctionDef) ctx.syntaxNode();
+    Set<Tree> secondaryLocations = Optional.ofNullable(functionDef.parameters())
+      .map(ParameterList::nonTuple)
+      .stream()
+      .flatMap(List::stream)
+      .map(Parameter::typeAnnotation)
+      .filter(Objects::nonNull)
+      .map(TypeAnnotation::expression)
+      .filter(GenericFunctionTypeParameterCheck::isGenericTypeAnnotation)
+      .collect(Collectors.toSet());
+    Optional.ofNullable(functionDef.returnTypeAnnotation())
+      .map(TypeAnnotation::expression)
+      .filter(GenericFunctionTypeParameterCheck::isGenericTypeAnnotation)
+      .ifPresent(secondaryLocations::add);
+
+    if (!secondaryLocations.isEmpty()) {
+      PreciseIssue issue = ctx.addIssue(functionDef.name(), MESSAGE);
+      secondaryLocations.forEach(loc -> issue.secondary(loc, SECONDARY_MESSAGE_USE));
+      getAssignmentLocations(secondaryLocations).forEach(loc -> issue.secondary(loc, SECONDARY_MESSAGE_ASSIGNMENT));
+    }
+  }
+
+  private static Set<Tree> getAssignmentLocations(Set<Tree> secondaryLocations) {
```

Comment:
It is a bit unfortunate to recompute the assigned value location when it's already done in `isGenericTypeAnnotation`. 

I have the feeling that method could probably return a list of `IssueLocation` that would be ready to add on the issue (it would then have to distinguish explicitly between the parameter/assigned value and to ensure no duplicate gets through - and probably would be easier to code imperatively than with the stream).

That said, I think it's okay to keep as it is now, since it won't change much in the grand scheme of things.

Reply:
Thanks for the tip.